test-app/tests/integration/components/checkbox-field-test.gts

```diff
@@ -0,0 +1,223 @@
+      .hasClass('shadow-error-outline');
+  });
+
+  test('it sets the "for" attribute on the label to the "id" attribute of the checkbox', async function (assert) {
+    await render(<template>
+      <CheckboxField @label="Label" data-checkbox />
+    </template>);
+
+    let labelFor = find('[data-control] > label')?.getAttribute('for') || '';
+    assert.ok(labelFor, 'Expected the id attribute of the label to be truthy');
+
+    let checkboxId = find('[data-checkbox]')?.getAttribute('id') || '';
+    assert.strictEqual(
+      checkboxId,
+      labelFor,
+      'Expected the for attribute on the label to match the id attribute on the checkbox'
+    );
+  });
+
+  test('it disables the checkbox using `@isDisabled`', async function (assert) {
+    await render(<template>
+      <CheckboxField @label="Label" @isDisabled={{true}} data-checkbox />
+    </template>);
+
+    assert.dom('[data-checkbox]').isDisabled();
+  });
+
+  test('it spreads attributes to the underlying checkbox', async function (assert) {
+    await render(<template>
+      <CheckboxField @label="Label" name="checkbox-name" data-checkbox />
+    </template>);
+
+    assert.dom('[data-checkbox]').hasAttribute('name', 'checkbox-name');
+  });
+
+  test('it sets the checked-state via `@value`', async function (assert) {
+    await render(<template>
+      <CheckboxField @label="Label" @value={{true}} data-checkbox />
+    </template>);
+
+    assert.dom('[data-checkbox]').isChecked();
+  });
+
+  test('it calls `@onChange` when clicked with the expected checked state', async function (assert) {
+    assert.expect(8);
+
+    let handleChange = (
+      checked: boolean,
+      e: Event | InputEvent,
+      isIndeterminate: boolean
+    ) => {
+      assert.true(
+        checked,
+        'Expected to be checked since we started un-checked'
+      );
+      assert.ok(e, 'Expected `e` to be available as the second argument');
+      assert.ok(e.target, 'Expected direct access to target from `e`');
+      assert.false(
+        isIndeterminate,
+        'Expected indeterminate state to be false as we did not provide `@isIndeterminate={{true}}`'
+      );
+      assert.step('handleChange');
+    };
+
+    await render(<template>
+      <CheckboxField
+        @label="Label"
+        @onChange={{handleChange}}
+        @value={{false}}
+        data-checkbox
+      />
+    </template>);
+
+    assert.verifySteps([]);
+
+    await click('[data-checkbox]');
+
+    assert.verifySteps(['handleChange']);
+
+    assert.dom('[data-checkbox]').isChecked();
+  });
+
+  test('it calls `@onChange` when clicked with the expected indeterminate state', async function (assert) {
+    assert.expect(4);
+
+    let handleChange = (
+      _checked: boolean,
+      _e: Event | InputEvent,
+      isIndeterminate: boolean
+    ) => {
+      assert.false(
+        isIndeterminate,
+        'Expected indeterminate state to be false as we did not provide `@isIndeterminate={{true}}`'
```

Comment:
Didn't we provide `@isIndeterminate={{true}}` on line 184 (below)? Does clicking make `Indeterminate` false?

Reply:
I always thought `isIndeteminate` is for bulk selection...

Reply:
I agree it's a bit confusing. I believe what is happening is that by clicking the checkbox, the browser itself sets `element.indeterminate` to false. And `@onChange` is _not_ called with what is passed as `@isIndeterminate`, but with the actual property value of the input element (which again has changed). That's why we receive false, while still true is passed as `@isIndeterminate`. Am I understanding this right?

But then the assertion message is wrong. Should it be something like this?

```suggestion
        'Expected indeterminate state to be false as the indeterminate property on the input element was changed by clicking it`'
```

Reply:
Okay thanks this clarifies this for me.